drivers/hmis/app/models/hmis/user.rb

```diff
@@ -88,6 +88,11 @@ def load_effective_permissions
       joins(:roles).
         where(roles: { permission => true })
     end
+
+    scope "#{permission}_for", ->(_entity) do
+      # todo @martha - we want to query for only users that have permission for this entity, but it's complicated
+      joins(:roles).where(roles: { permission => true }).group(:id)
```

Comment:
it might be worth asking Elliot about this, we could well have something already implemented on the warehouse side, which uses similar models

Reply:
@eanders , do you know if we do anything similar to this in the warehouse? From what I could find, not yet:
https://github.com/greenriver/hmis-warehouse/blob/9755f5b77568378ef4f160713db8579848c4d63d/app/models/user.rb#L73-L78
is the most similar, but doesn't include a scope that checks for all users with permission *for* a particular entity.

Reply:
We usually do the check on the entity `scope :viewable_by, -> (user)`.  (See `app/models/grda_warehouse/hud/project.rb#viewable_by` for one example.)

The trick with going the other way is that you have to walk the inheritance chain, users can inherit access to a project from the project's organization for instance.

I think this would be something like (untested code):
```
scope "#{permission}_for", ->(entity) do
 user_ids = Hmis::AccessControl.joins(:role, :access_group, user_group: :users).
    preload(user_group: :user_group_members).
    merge(Hmis::Role.where(permission => true)).
    merge(Hmis::AccessGroup.contains(entity)). # I think this `contains` scope isn't complete since it doesn't walk the inheritance tree (but maybe it should?)
    select(Hmis::User.arel_table[:id]) # selecting the ids so we can ensure the returned scope doesn't include any of the complexity 
  User.where(id: user_ids)
end

```

Looking through the code, I think it would be helpful to add a `contains_with_inherited` scope on `Hmis::AccessGroup`.  That will be a bit of a pain to write, but should make the above useful.

Reply:
Thanks @eanders this is extremely helpful. Does this look like I am on the right track?
https://github.com/greenriver/hmis-warehouse/pull/4564/files#diff-c4a75867ebfe78fba438a93af5a60cddcd44463aaa2be29eac798cc3902d30a0R49-R62

- Do you think it is OK to leave the other entity types as 'not implemented' for now until a use case arises?

Reply:
Hmm, so I think I must not have been very clear.  (Happy to huddle about it if that'd be easier.)  What you have still doesn't include the inheritance (as far as I can see.)  When checking for projects, you need to include the organizations and data sources that also contain that project since permission may be granted further up the chain.  

I'll drop a comment on that scope to see if I can get closer.